### PR TITLE
clang-tidy: modernize-use-using respect externC

### PR DESCRIPTION
**Pull request by dkalowsk**
First pass at fixing the modernize-use-using to respect pieces marked as externC.  

This implementation passes all local tests, but still fails when running on Esri code.  Details in a comment below

### CODE REVIEW

**Comment by dkalowsk** [comment]
It appears that clang doesn't always respect the:
```
#ifdef _cplusplus
extern C {
#endif
```
linkage change.  For example, running the clang-check on the unit test for use-using:
```
~/clang/esri/bin/clang-check -ast-dump -extra-arg="-std=c++14" modernize-use-using.cpp 

/* ... lots of extra AST here */
`-LinkageSpecDecl 0x7facef86cae8 <line:167:1, line:181:1> line:167:8 C
  |-CXXRecordDecl 0x7facef86cb38 <line:170:1, col:8> col:8 struct _RT_Struct
  |-TypedefDecl 0x7facef86cca0 <line:171:1, col:28> col:28 RT_ExtCHandle 'struct _RT_Struct *'
  | `-PointerType 0x7facef86cc50 'struct _RT_Struct *'
  |   `-ElaboratedType 0x7facef86cbf0 'struct _RT_Struct' sugar
  |     `-RecordType 0x7facef86cbd0 'struct _RT_Struct'
  |       `-CXXRecord 0x7facef86cb38 '_RT_Struct'
  |-EnumDecl 0x7facef86ccf8 <line:173:9, line:178:1> line:173:14 RT_EnumValue
  | |-EnumConstantDecl 0x7facef86cdd0 <line:175:3, col:16> col:3 RT_EnumZero 'enum RT_EnumValue'
  | | `-ImplicitCastExpr 0x7facef86cee8 <col:16> 'unsigned int' <IntegralCast>
  | |   `-IntegerLiteral 0x7facef86cdb0 <col:16> 'int' 0
  | |-EnumConstantDecl 0x7facef86ce38 <line:176:3, col:15> col:3 RT_EnumOne 'enum RT_EnumValue'
  | | `-ImplicitCastExpr 0x7facef86cf00 <col:15> 'unsigned int' <IntegralCast>
  | |   `-IntegerLiteral 0x7facef86ce18 <col:15> 'int' 1
  | `-EnumConstantDecl 0x7facef86cea0 <line:177:3, col:15> col:3 RT_EnumTwo 'enum RT_EnumValue'
  |   `-ImplicitCastExpr 0x7facef86cf18 <col:15> 'unsigned int' <IntegralCast>
  |     `-IntegerLiteral 0x7facef86ce80 <col:15> 'int' 2
  `-TypedefDecl 0x7facef86cf80 <line:173:1, line:178:3> col:3 RT_EnumExternalValues 'enum RT_EnumValue':'enum RT_EnumValue'
    `-ElaboratedType 0x7facef86cf30 'enum RT_EnumValue' sugar
      `-EnumType 0x7facef86cd90 'enum RT_EnumValue'
        `-Enum 0x7facef86ccf8 'RT_EnumValue'
```
We clearly see the added test code is using the LinkageSpecDecl defined in C.  

From our own headers we would expect to see the same entry at the very top, but... 
```
TranslationUnitDecl 0x7fc9528318d0 <<invalid sloc>> <invalid sloc>
|-TypedefDecl 0x7fc952831e60 <<invalid sloc>> <invalid sloc> implicit __int128_t '__int128'
| `-BuiltinType 0x7fc952831b40 '__int128'
|-TypedefDecl 0x7fc952831ed0 <<invalid sloc>> <invalid sloc> implicit __uint128_t 'unsigned __int128'
| `-BuiltinType 0x7fc952831b60 'unsigned __int128'
|-TypedefDecl 0x7fc952832218 <<invalid sloc>> <invalid sloc> implicit __NSConstantString 'struct __NSConstantString_tag'
| `-RecordType 0x7fc952831fc0 'struct __NSConstantString_tag'
|   `-CXXRecord 0x7fc952831f28 '__NSConstantString_tag'
|-TypedefDecl 0x7fc9528322b0 <<invalid sloc>> <invalid sloc> implicit __builtin_ms_va_list 'char *'
| `-PointerType 0x7fc952832270 'char *'
|   `-BuiltinType 0x7fc952831960 'char'
|-TypedefDecl 0x7fc952868400 <<invalid sloc>> <invalid sloc> implicit __builtin_va_list 'struct __va_list_tag [1]'
| `-ConstantArrayType 0x7fc952832590 'struct __va_list_tag [1]' 1 
|   `-RecordType 0x7fc9528323a0 'struct __va_list_tag'
|     `-CXXRecord 0x7fc952832308 '__va_list_tag'
|-CXXRecordDecl 0x7fc952868488 </Users/dkalowsky/development/rtc/runtimecore/c_api/include/c_api/mapping/symbol/labeling/labeling_types.h:25:1, col:8> col:8 struct _RT_LabelDefinitionInternal
|-TypedefDecl 0x7fc9528685f0 <line:26:1, col:45> col:45 RT_LabelDefinitionHandle 'struct _RT_LabelDefinitionInternal *'
```

Note the lack of any Linkage definition.

**Comment by dkalowsk** [comment]
Okay, nevermind the problems, it is now proving to work.  The fix was done internally.  Review these changes and let's move on.

**Comment by dkalowsk** [comment]
The question is, do we try to push this upstream?

**Comment by chri7325** [comment]
I think that would be really good.  @duncanthomson opened an issue for it.  Since this repo is not directly connected to the llvm source control (I think they use cvs), it might be good enough to just copy paste the patch into the issue so they could pick it up.

File: clang-tools-extra/clang-tidy/modernize/UseUsingCheck.cpp [hunk 1]
```diff
@@ -83,6 +83,20 @@ void UseUsingCheck::check(const MatchFinder::MatchResult &Result) {
...
+      }
+    }
+  }
+
+  if (const auto *ED = dyn_cast<EnumDecl>(MatchedDecl)) {
+    if (ED->isExternCContext()) {
+      return;
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (dkalowsk) [reply clang-tools-extra/clang-tidy/modernize/UseUsingCheck.cpp]: Looks like this suggested fix works as well.  Nice change.  It's simpler and cleaner.  I'll push up the change in a moment.  

Regarding adding to the `TypedefDecl`, I'm fairly certain that might be the correct solution.  I'm not convinced that we want to go that route though, as it has a lot of more subtle impacts that I'm not sure we're going to be aware of.  If you can get upstream clang to take it though I'm happy to revert out these changes :-)

**Comment by chri7325** [comment]
@duncanthomson now you can push ;)

**Pull request merged**